### ISSUE TITLE
normalizeNodeAndChildren does not rsync node with change object at enough intervals

### ISSUE THREAD

**Issue by CameronAckermanSEL** [ianstormtaylor/slate#1528]
#### Do you want to request a *feature* or report a *bug*?

Bug

#### What's the current behavior?

With certain very complicated schema rules, normalizeNodeAndChildren can become badly out of sync with the current document state, even though there are some safeguards against this sort of situation.

The root cause is the code here:
https://github.com/ianstormtaylor/slate/blob/5444a300b8fe86edbe2250da86f11e8a9d0ae776/packages/slate/src/changes/with-schema.js#L86
```
    // Unwind the stack, normalizing every child and adding it to the set.
    while (key = keys[0]) {
      const child = node.getChild(key)
      normalizeNodeAndChildren(change, child, schema)
      set.add(key)
      keys.shift()
    }
```

Here the normalization will recurse into child blocks of the node being validated. Now, if the recursive calls here spawn normalization that causes the descendant structure of the node to change, then the node is out of sync of the change object within this loop. Children of this node are no longer guaranteed to be in sync with the change object, and may not even exist within the node under validation.

Changing the code to this resolves the issue:

```
function normalizeNodeAndChildren(change, node, schema) {
  if (node.object == 'text') {
    normalizeNode(change, node, schema)
    return
  }

  // We can't just loop the children and normalize them, because in the process
  // of normalizing one child, we might end up creating another. Instead, we
  // have to normalize one at a time, and check for new children along the way.
  // PERF: use a mutable array here instead of an immutable stack.
  const keys = node.nodes.toArray().map(n => n.key)

  let { size } = change.operations
  let key

  // PERF: use a mutable set here since we'll be add to it a lot.
  let set = new Set()

  // Unwind the stack, normalizing every child and adding it to the set.
  while (key = keys[0]) {
    set = set.add(key)
    keys.shift()
    const child = node.getChild(key)
    if (child) {
      normalizeNodeAndChildren(change, child, schema)
      // PERF: Only re-find the node and re-normalize any new children if
      // operations occured that might have changed it.
      if (change.operations.size > size) {
        size = change.operations.size
        node = refindNode(change, node)
        // if the node is no longer in the document
        // then break the loop, dont need to normalize
        // it anymore
        if (!node) break

        // Add any new children back onto the stack.
        node.nodes.forEach((n) => {
          if (set.has(n.key)) return
          keys.unshift(n.key)
        })
      }
    }
  }

  // Normalize the node itself if it still exists.
  if (node) {
    normalizeNode(change, node, schema)
  }
}
```

The setup that resulted in this validation problem was
1. Schema has a rule that flattens nested lists because our system can't support it yet (sadface)
2. User pastes text fragment that includes a list item from a list that is nested within another list (so list item depth 2), and then the following block is a paragraph.

Chain of events is then:
1. Validation will start at the document level, then iterate over all children recursively.
2. Eventually, validation will visit the nested list and promote it to the document level through several validation rules.
3. This in turn triggers another validation pass, which results in the validation of two adjacent text nodes in the final paragraph of the paste (insertFragment creates this condition). The two text nodes are merged correctly without any problems.
4. The recursive validation ends here, then flows back up to the document level.
5. The loop in the code snippet visits the paragraph block that _had_ adjacent text nodes. In the change object, this condition is resolved, but because the document node has not been resynced, the document node still has the text node rule violation.
6. The validate node function that merges text nodes fires and blows up because the change object no longer has this rule violation and can't find the second text node for merging.
7. Exception occurs and paste fails.

Unfortunately the code I have is closed source and it is very hard to reproduce in fiddle =\

I can work on providing a sample of code that reproduces this  defect if necessary, but it will take a lot of extra time.

#### What's the expected behavior?

normalization functions are synchronized well with the current document change value while maintaining adequate performance.

**Comment by CameronAckermanSEL** [comment]
Willing to submit a PR, but just want to make sure we got this right and you're ok with our proposal. :)

**Comment by DamareYoh** [comment]
Fixed in #1538 , thanks @zhujinxuan !
